src/kaleidoscope/hardware/kbdfans/KBD4x.h

```diff
@@ -21,35 +21,42 @@
 
 #include <Arduino.h>
 #define HARDWARE_IMPLEMENTATION kaleidoscope::hardware::kbdfans::KBD4x
-#include "Kaleidoscope-HIDAdaptor-KeyboardioHID.h"
 
-#include "kaleidoscope/macro_helpers.h"
-#include "kaleidoscope/hardware/avr/pins_and_ports.h"
+#include "kaleidoscope/hardware/kbdfans/kbd4x/KBD4xScannerDescription.h"
 
-#include "kaleidoscope/hardware/ATMegaKeyboard.h"
+#include "kaleidoscope/driver/led/NoLeds.h"
+#include "kaleidoscope/driver/mcu/AVR.h"
+#include "kaleidoscope/driver/scanner/ATMegaScanner.h"
+
+#include "kaleidoscope/HardwareDescription.h"
+#include "kaleidoscope/Hardware.h"
 
 namespace kaleidoscope {
 namespace hardware {
 namespace kbdfans {
-class KBD4x: public kaleidoscope::hardware::ATMegaKeyboard {
+
+using KBD4xScanner = kaleidoscope::driver::scanner::ATMegaScanner<KBD4xScannerDescription>;
+
+struct KBD4xDescription : public kaleidoscope::HardwareDescription {
+  typedef KBD4xScannerDescription ScannerDescription;
+  typedef KBD4xScanner KeyScanner;
+  typedef kaleidoscope::driver::mcu::AVR MCU;
+};
+
+class KBD4x: public kaleidoscope::Hardware<KBD4xDescription> {
  public:
   KBD4x(void) {
     mcu_.disableJTAG();
     mcu_.disableClockDivision();
   }
 
-  ATMEGA_KEYBOARD_CONFIG(
-    ROW_PIN_LIST({ PIN_D0, PIN_D1, PIN_D2, PIN_D3 }),
-    COL_PIN_LIST({ PIN_F0, PIN_F1, PIN_F4, PIN_F5, PIN_F6, PIN_F7, PIN_B3, PIN_B1, PIN_B0, PIN_D5, PIN_B7, PIN_C7 })
-  );
-
   static constexpr int8_t led_count = 0;
 
-  static constexpr int8_t numKeys() {
-    return matrix_columns * matrix_rows;
-  }
-
   void resetDevice();
+
+  void setup() {
```

Comment:
Wouldn't it make sense to let the `Hardware` template base class have a `setup()` method that does the default job? It can be reimplemented emptily if necessary? Then, all component could also have a - possibly empty and inlined - `setup()` method that would be called by default from the `Hardware` base template. 

To add additional setup, one would write
```cpp
class DerivedHardware : public Hardware<...> {
   ...
   typedef Hardware<...> BaseHardware;

   void setup() {
      BaseHardware::setup();
      ...additional setup code
   }
   ...
};